AArch64: Implementation of ArrayLengthEvaluator

marufunb opened the pull request:
This commit implements the arraylengthevaluator for openj9.

Signed-off-by: marufunb <mrahma15@unb.ca>

marufunb commented:
@0xdaryl If you can please manage sometime to look into my implementation of arraylengthevaluator, it will be really helpful for me.

charliegracie commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -39,3 +39,24 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
+   TR::Register *lengthReg = cg->allocateRegister();
+
+   TR::MemoryReference *contiguousArraySizeMR = new (cg->trHeapMemory()) TR::MemoryReference(objectReg, fej9->getOffsetOfContiguousArraySizeField(), 4, cg);
+   TR::MemoryReference *discontiguousArraySizeMR = new (cg->trHeapMemory()) TR::MemoryReference(objectReg, fej9->getOffsetOfDiscontiguousArraySizeField(), 4, cg);
```
where is discontiguousArraySizeMR used?

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -39,3 +39,24 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
+   TR::MemoryReference *discontiguousArraySizeMR = new (cg->trHeapMemory()) TR::MemoryReference(objectReg, fej9->getOffsetOfDiscontiguousArraySizeField(), 4, cg);
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, lengthReg, contiguousArraySizeMR);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::ccmpimmw, node, lengthReg, lengthReg);
```
This is an immediate-form instruction.  You're missing the immediate.  Likely "0".   Please look into whether a TST instruction would be a better choice here.

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -39,3 +39,24 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, lengthReg, contiguousArraySizeMR);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::ccmpimmw, node, lengthReg, lengthReg);
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, lengthReg, discontiguousArraySizeMR);
```
This instruction simply overwrites the `lengthReg` you set previously.  I think what you want is some variant of `CSEL`.

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -39,3 +39,24 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->fe());
+   // MOV R, [B + contiguousSize]
+   // TEST R, R
+   // CMOVE R, [B + discontiguousSize]
```
These are x86-style comments.  Please update for AArch64.

marufunb commented:
@0xdaryl I have updated the implementation according to your feedback. Can you please manage sometime to look at that?

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -39,3 +39,31 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, discontiguousArrayLabel, condReg, TR::CC_EQ);
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, doneLabel);
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, lengthReg, discontiguousArraySizeMR);
+   generateLabelInstruction(cg, TR::InstOpCode::b, node, doneLabel);
```
You have to be careful about introducing control flow in register sequences because of the interaction with the register assigner (which operates linearly).  You don't want spilling to occur on one path and not the other.  Generally, the way you work around that is to force spilling at a controlled point, namely the merge point label.  The way you do that is to add a `NoReg` register dependency to this label for the registers being referenced (in your case `lengthReg`).

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -39,3 +39,31 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
+   // ldrw R, [B + contiguousSize]        ; Load contiguous array length
+   // cmp R, 0                            ; If 0, must be a discontiguous array
+   // ldrw R, [B + discontiguousSize]     ; Load discontiguous array length
+   //b done
```
This comment doesn't match the code.

marufunb commented:
@0xdaryl Can you please manage sometime to have a look to this updated one?

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -39,3 +39,27 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, lengthReg, contiguousArraySizeMR);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::ccmpimmw, node, condReg, lengthReg, 0);
+   generateTrg1MemInstruction(cg, TR::InstOpCode::cselw, node, lengthReg, discontiguousArraySizeMR);
```
Please read the specification for the CSEL instruction (section C6.2.61 in the ARM Architecture Reference Manual for ARMv8-A) for its format and semantics.

You'll find that this instruction actually operates on 3 registers and a condition:
```
   CSEL <Xd>, <Xn>, <Xm>, <cond>
```
and who's operation is defined (very loosely) as:
```
   Xd = cond ? Xn : Xm
```
This means you'll have to load the discontiguous array length into a register first.

Looking into this deeper, I don't believe we have an instruction class for this kind of instruction yet, so you'll have to create one in OMR.  I think creating a `ARM64CondTrg1Src2Instruction` that extends `ARM64Trg1SrcInstruction` is the way to go.  One of the constructor APIs for such an instruction should look like this:
```
   ARM64CondTrg1Src2Instruction(
         TR::InstOpCode::Mnemonic op,
         TR::Node *node,
         TR::ARM64ConditionCode cc,
         TR::Register *treg,
         TR::Register *s1reg,
         TR::Register *s2reg,
         TR::CodeGenerator *cg) :
      ARM64Trg1Src2Instruction(op, node, treg, s1reg, s2reg, cg),
         _conditionCode(cc)
      {
      ...
      }
   ```
Please have a look at creating a new instruction kind.  You can probably follow the pattern of other instructions.

@knn-k in case he has any different opinions on this.

knn-k replied:
I agree with @0xdaryl on making a new Instruction class based on existing `ARM64Trg1Src2Instruction` plus `ARM64ConditionCode`.
Please look at `ARM64Trg1Src2ShiftedInstruction` for how to add an extra element to `ARM64Trg1Src2Instruction`.

aarongraham9 commented:
What's the status of this one @marufunb?

marufunb commented:
It depends on the instruction I implemented for OMR. I have updated it but did not push to the PR.

Thanks,
Md. Mahbubur Rahman (Maruf)
Graduate Academic Assistant at
IBM/UNB Centre of Advanced Studies- Atlantic (CASA)
Phone: 5064714698
Faculty of Computer Science,
University of New Brunswick

________________________________
From: Aaron Graham <notifications@github.com>
Sent: Tuesday, July 2, 2019 4:41:57 PM
To: eclipse/openj9
Cc: Md. Mahbubur Rahman; Mention
Subject: Re: [eclipse/openj9] WIP Implementation of ArrayLengthEvaluator: (#6077)


✉External message: use caution.

What's the status of this one @marufunb<https://github.com/marufunb>?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/eclipse/openj9/pull/6077?email_source=notifications&email_token=AKOJ5FLMNWZJZQDGFB4DEDDP5OVQLA5CNFSM4HWUMJIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZCK6LI#issuecomment-507817773>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKOJ5FOX2CHNZNYPJVHTKFTP5OVQLANCNFSM4HWUMJIA>.

marufunb commented:
@0xdaryl I have made those changes and pushed the updated one. When you can manage time please have a look at this updated one.

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -39,3 +39,30 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
+   //
+   TR::Register *objectReg = cg->evaluate(node->getFirstChild());
+   TR::Register *lengthReg = cg->allocateRegister();
+   TR::Register *condReg = cg->allocateRegister(TR_CCR);
```
AArch64 does not have the notion of a condition code register.

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -39,3 +39,30 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
+   TR::Register *discontiguousLengthReg = cg->allocateRegister();
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, lengthReg, contiguousArraySizeMR);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::ccmpimmw, node, condReg, lengthReg, 0);
```
This is neither the right form of instruction to use nor the right opcode.  Please use a `generateCompareImmInstruction` of the `lengthReg` to `0`.

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -39,3 +39,30 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
+
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, lengthReg, contiguousArraySizeMR);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::ccmpimmw, node, condReg, lengthReg, 0);
+   generateCondTrg1Src2Instruction(cg, TR::InstOpCode::cselw, node, lengthReg, lengthReg, discontiguousLengthReg, TR::CC_EQ);
```
You will need to load the discontiguous length field into a register (i.e., `discontiguousLengthReg`) prior to this instruction (like you did for the contiguous length field).

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -39,3 +39,30 @@ J9::ARM64::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::ccmpimmw, node, condReg, lengthReg, 0);
+   generateCondTrg1Src2Instruction(cg, TR::InstOpCode::cselw, node, lengthReg, lengthReg, discontiguousLengthReg, TR::CC_EQ);
+
+   cg->stopUsingRegister(condReg);
```
Not needed.

marufunb commented:
@0xdaryl I have made those changes. Please have a look when you have sometime.

knn-k commented:
You have merge conflicts with #6055.

0xdaryl commented:
@marufunb , please resolve the merge conflicts.  Can you confirm that this PR builds on AArch64 with the needed changes from OMR?

marufunb commented:
> @marufunb , please resolve the merge conflicts. Can you confirm that this PR builds on AArch64 with the needed changes from OMR?

Yes hopefully I will resolve the merge conflict and build it following Aaron's documentation today asap.

0xdaryl commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -136,3 +136,30 @@ J9::ARM64::TreeEvaluator::checkcastEvaluator(TR::Node *node, TR::CodeGenerator *
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, lengthReg, contiguousArraySizeMR);
+   generateCompareImmInstruction(cg, node, lengthReg, 0);
+   generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, discontiguousLengthReg, discontiguousArraySizeMR);
+   generateCondTrg1Src2Instruction(cg, TR::InstOpCode::cselw, node, lengthReg, lengthReg, discontiguousLengthReg, TR::CC_EQ);
```
I think the condition is backwards.  You want `TR::CC_NE` if you're using the registers in the `CSEL` in that order.  If the arraylength is zero, its true length is in the discontiguous length field.

0xdaryl commented:
Almost there!  Needs a minor fix and a rebase.

marufunb commented:
@0xdaryl I have fixed that and rebased. Please have a look when you get sometime.

knn-k commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -148,6 +148,7 @@ J9::ARM64::TreeEvaluator::checkcastEvaluator(TR::Node *node, TR::CodeGenerator *
    }
 
 TR::Register *
+<<<<<<< HEAD
```
The line `<<<<<<< HEAD` must be removed.

knn-k commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -157,3 +158,30 @@ J9::ARM64::TreeEvaluator::newObjectEvaluator(TR::Node *node, TR::CodeGenerator *
    return targetRegister;
    }
 
+=======
```
This line `=======` and the line `>>>>>>> AArch64: Implementation of ArrayLengthEvaluator` at the end of the function must be removed, too.

marufunb commented:
@0xdaryl @knn-k Please  have a look when you have time. I think now I have resolved all the requested changes.

aarongraham9 commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -187,3 +187,29 @@ J9::ARM64::TreeEvaluator::monentEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->fe());
+   // ldrw R, [B + contiguousSize]        
+   // cmp R, 0                            
+   // cselw R, [B + discontiguousSize]    
```
Are these the `AArch64` instructions/pseudo-code for the operation? If not are the comments necessary?

marufunb replied:
Yes I wrote those for AArch64.

aarongraham9 commented on runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:
```diff
@@ -187,3 +187,29 @@ J9::ARM64::TreeEvaluator::monentEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   TR::Register *discontiguousLengthReg = cg->allocateRegister();
+
+   TR::MemoryReference *contiguousArraySizeMR = new (cg->trHeapMemory()) TR::MemoryReference(objectReg, fej9->getOffsetOfContiguousArraySizeField(), 4, cg);
+   TR::MemoryReference *discontiguousArraySizeMR = new (cg->trHeapMemory()) TR::MemoryReference(objectReg, fej9->getOffsetOfDiscontiguousArraySizeField(), 4, cg);   
```
Shouldn't the discontiguous case memory size only be calculated if this is a discontiguous array? Same for the contiguous case? Can you check this beforehand?

Do we care about discontiguous arrays/arraylets at this point?

marufunb replied:
Yes we care about the discontiguous arrays/arraylets. I am not sure whether we can check that beforehand or not because we are comparing the calculated memory size with 0 to differentiate between contiguous and discontiguous in the later instructions, and I have followed the structure of Power and X.

0xdaryl commented:
Jenkins build is not required.  Merging.

knn-k commented:
This PR caused a build break because we don't have the `MemoryReference` constructor with arguments `(TR::Register *, offset, size, TR::CodeGenerator *)` for AArch64.

https://github.com/eclipse/openj9/blob/4eb20ee6d69fcca8e2eb3b0f091d44648859f5f9/runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp#L202

0xdaryl commented:
I think dropping the "4" parameter is what you want.  @marufunb would you mind fixing please.  @knn-k, if you need a fix in the next 45 minutes I can review/merge.

knn-k commented:
I opened #6407 for fixing the build break.
